### PR TITLE
RevEng: Custom Templates

### PR DESCRIPTION
**Pull request by lajones**
Addresses issues #2415, #2418 & #2581.
Allow outputting of default templates to user's project directory.
Use such templates instead of the defaults if they are available.
Various tab-expansion improvements for Powershell version of the commands.

### CODE REVIEW

File: src/EntityFramework.Commands/tools/EntityFramework.psm1 [hunk 1]
```diff
@@ -332,6 +332,57 @@ function Reverse-Engineer {
...
+    Project = { GetProjects }
+    StartupProject = { GetProjects }
+    Provider = {
+        param ($tabExpansionContext)
+
+        $projectName = $tabExpansionContext.Project
+        if (!($projectName)) {
+            $projectName = (Get-Project).ProjectName
+        }
+        return Get-Package -ProjectName $projectName | select -ExpandProperty Id
```
Comment (bricelam) [inline src/EntityFramework.Commands/tools/EntityFramework.psm1]: DRY. Pull this out into its own method now.

File: src/EntityFramework.Commands/tools/EntityFramework.psm1 [hunk 2]
```diff
@@ -332,6 +332,57 @@ function Reverse-Engineer {
...
+.DESCRIPTION
+	Outputs templates which can then be customized by the user to control how to reverse engineer code from a database.
+
+.PARAMETER Provider
+	Specifies the name of the provider for which to output the templates, for example EntityFramework7.SqlServer
+
+.PARAMETER Project
+	Specifies the project to use. If omitted, the default project is used.
+
+.PARAMETER $StartupProject
```
Comment (bricelam) [inline src/EntityFramework.Commands/tools/EntityFramework.psm1]: Oops, did I put that `$` there?

File: src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringGenerator.cs [hunk 3]
```diff
@@ -200,5 +233,17 @@ private static void CheckConfiguration(ReverseEngineeringConfiguration configura
                 throw new ArgumentException(Strings.NamespaceRequired);
             }
         }
+
+        private string LoadTemplate(string searchPath, string fileName, Func<string> providerTemplateFunc)
+        {
+            if (!string.IsNullOrEmpty(searchPath)
+                && FileService.FileExists(searchPath, fileName))
+            {
+                Logger.LogInformation(Strings.UsingCustomTemplate(searchPath + @"\" + fileName));
```
Comment (bricelam) [inline src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringGenerator.cs]: `Path.Combine()`?

**Comment by bricelam** [comment]
:shipit:

File: src/EntityFramework.Commands/tools/EntityFramework.psm1 (hunk 1 above)
Comment (lajones) [inline src/EntityFramework.Commands/tools/EntityFramework.psm1]: Will do.

File: src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringGenerator.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringGenerator.cs]: Will do.

File: src/EntityFramework.Commands/tools/EntityFramework.psm1 (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.Commands/tools/EntityFramework.psm1]: Yes. But I copied it around without noticing. Erik spotted it.

**Comment by lajones** [comment]
Checked in - commit 5f02ee3.

**Pull request closed without merging**